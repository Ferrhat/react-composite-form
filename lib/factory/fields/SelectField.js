import React, {Component} from 'react';
import Select from 'react-select';
import { get, isNumber, orderBy, defaultTo } from 'lodash';

class SelectField extends Component {
    render() {
        const { name, label, value, valueUpdateHandler, changeHandler, disabled, options, labelKey, valueKey, placeholder, staticField, sortBy } = this.props;

        let transformedOptions = options.map(option => ({label: option[labelKey], value: option[valueKey]}));

        const onChange = (selection) => {
            valueUpdateHandler(name, get(selection, 'value', null));
            changeHandler(selection);
        }

        const sortProperty = get(sortBy, 'propertyName', '');
        const sortOrder = get(sortBy, 'order', 'asc');
        if (sortProperty) {
            transformedOptions = orderBy(transformedOptions, sortProperty, sortOrder);
        } else {
            transformedOptions = orderBy(transformedOptions, 'label');
        }

        return (
            <div>
                <Select
                    value={ defaultTo(value, null) }
                    onChange={ onChange }
                    options={ transformedOptions }
                    isLoading={!transformedOptions.length && !staticField}
                    disabled={disabled || !transformedOptions.length}
                    placeholder={placeholder}
                />
                <div className={'validationErrorMessage'}>{get(this.props.errors, '0.message', '')}</div>
            </div>
        );
    }
}

SelectField.defaultProps = {
    labelKey: 'label',
    valueKey: 'value',
};

export default SelectField;
