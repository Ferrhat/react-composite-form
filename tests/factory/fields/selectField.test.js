import React from 'react';
import {shallow, mount} from 'enzyme';
import SelectField from "../../../lib/factory/fields/SelectField";


describe('SelectField', () => {
    const mockValueUpdateHandler = jest.fn();
    const mockChangeHandler = jest.fn();
    let selectField = shallow(<SelectField valueUpdateHandler={mockValueUpdateHandler} changeHandler={mockChangeHandler} options={[]} name={'testName'} />);

    it('renders properly', () => {
        expect(selectField).toMatchSnapshot();
    });

    it('contains a connected Select component', () => {
        expect(selectField.find('Select').length).toEqual(1);
    });

    it('should trigger a change event', () => {
        const selection = {
            value: 'testValue',
        };
        selectField.find('Select').simulate('change', selection);
        expect(mockValueUpdateHandler).toBeCalledWith('testName', 'testValue');
        expect(mockChangeHandler).toBeCalledWith(selection);
    });

    it('stops loading if it contains options', () => {
        expect(selectField.find('Select').prop('isLoading')).toEqual(true);
        selectField.setProps({options: [{label: 'Test', value: 1}]}, () => {
            expect(selectField.find('Select').prop('isLoading')).toEqual(false);
        });
    });

    it('changes the id, name to label, value', () => {
        selectField = mount(<SelectField valueUpdateHandler={mockValueUpdateHandler} changeHandler={mockChangeHandler} options={[]} name={'testName'} />);
        selectField.setProps({options: [{name: 'Test', id: 1}], labelKey: 'name', valueKey: 'id'}, () => {
            expect(selectField.prop('options')).toEqual([{name: 'Test', id: 1}]);
            expect(selectField.find('Select').prop('options')).toEqual([{label: 'Test', value: 1}]);
        });
    });

});
