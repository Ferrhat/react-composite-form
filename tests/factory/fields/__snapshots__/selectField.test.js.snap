// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField renders properly 1`] = `
<div>
  <Select
    arrowRenderer={[Function]}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={true}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={true}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={true}
    joinValues={false}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={false}
    noResultsText="No results found"
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    options={Array []}
    pageSize={5}
    placeholder="Select..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    value={null}
    valueComponent={[Function]}
    valueKey="value"
  />
  <div
    className="validationErrorMessage"
  />
</div>
`;
